### DISCUSSION TITLE
Define translations directly in a test

### DISCUSSION THREAD

**Discussion by TristanPouliquen** [symfony/symfony#45225] [Q&A]
Hello,

We created a factory that generates responses with a certain type of error codes & a message that is automatically translated in our factory.

To test it, we want to dynamically add a default translation to our translator, call our factory and assert that the response contains the good error code & error message.

We tried 
```
$translator->addLoader('array', new ArrayLoader());
        $translator->addResource('array', [
            'hello' => 'Hello |name|!'
        ], 'fr_FR', 'api_response');
```

But the default `$translator` that we get from our container has the built-in cache system, which means that our new translation is not taken into account when the call to `Translator::trans()` is made later on.

How should we go about dynamically adding a translation to the translator?

We want to use as much as possible functional tests based on the services of the container so as not to couple our tests with the knowledge of how is written the SUT.

**Comment by stof** [comment]
I suggest handling that with a decorator for the translator, in which you could register some translation overrides (and then, any non-overriden message would be delegated to the original translator). And then, register that decorator in the `test` environment only (there is no point running it in prod, adding useless overhead).
